mrustc: new port

evanmiller opened the pull request:
#### Description

[mrustc](https://github.com/thepowersgang/mrustc) is an alternative Rust compiler suitable for bootstrapping Rust. The immediate motivation is to provide a basic Rust compiler on Snow Leopard, which isn't supported by the main `rust` port.

In the future `mrustc` might be used as a stage0 compiler for `rust`, but that's beyond the scope of this work.

`mrustc` requires a run-time C compiler. Since the system GCC on 10.6 cannot compile `mrustc`'s generated C files (which use thread-local storage), I have patched things to use `clang-mp-9.0` as the default run-time compiler on all platforms. This default can be overridden at run-time with `CC`. The choice of default compiler will be a good candidate for port variants in the future.

Tested through compiling and running a basic "Hello, world!" Rust program, so I think it's ready for others to play around with.
<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
-->
macOS 10.6.8
Xcode x.y

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

evanmiller commented:
> Installing dependency (9 of 25) 'xz' with variants '' ... [FAIL]

Anyone know why this fails on CI?

herbygillot commented:
> Anyone know why this fails on CI?

I've triggered the CI jobs to re-run.

evanmiller commented:
Updated the code and added a test target (which passes locally).

mascguy commented:
@catap Kirill, you might be interested in this too!

catap commented:
@mascguy thanks, it is interesting. I haven't got a time to finish 10.6 :) maybe someone else can do it.

@evanmiller FYI I've backported LLVM changes: https://reviews.llvm.org/D115250?id=392731 maybe you may suggest to use this patch at upstream of this tool

kencu commented:
My thread_local hack for llvm for <10.7 requires other changes in the built libc++abi to actually work, and should be considered "macports-only".

Upstream llvm will never accept them, as they have long dropped support for older darwin systems (the only reason older darwin works on macports is other patches and config hacks I made, and the stuff in legacysupport).

We should not be trying to upstream these hacks anywhere --just embarassing to do so, when people try them they won't work -- and consider them our mess to maintain while we do.

If you still try to upstream my tlv patches somewhere like llvm, please don't put my name on them!!!!

cjones051073 commented on lang/mrustc/Portfile:
```diff
@@ -0,0 +1,87 @@
+# This will be the default compiler, which can be overridden
+# at run-time with the CC environment variable.
+# Support variants in the future, but this will do for now
+depends_run-append port:clang-9.0
```
It would be better to not hardcode a version here, but to tweak the compiler backlisting etc. to ensure a compatible compiler is picked, and then to depend on that. This would allow newer compilers to be used when, e.g., clang-9.0 is not available (such as on newer systems which inevitably when one day stop supporting this compiler).

Note to add the correct runtime dep you can use the same logic as in rust.

https://github.com/macports/macports-ports/blob/a34b9f304f3a13bb9239e2ae9917ee3a41de3e9f/lang/rust/Portfile#L210

cjones051073 commented on lang/mrustc/Portfile:
```diff
@@ -0,0 +1,87 @@
+post-patch {
+    reinplace "s|@MACPORTS_MRUSTC_LIBDIR@|${prefix}/lib/mrustc-${rust_version_major}|" ${worksrcpath}/src/main.cpp
+    reinplace "s|@MACPORTS_MRUSTC_TARGET_VER@|${rust_version_major_underscore}|" ${worksrcpath}/src/main.cpp
+    reinplace "s|@MACPORTS_MRUSTC_CC@|clang-mp-9.0|" ${worksrcpath}/src/trans/codegen_c.cpp
```
once you have correctly set the compiler backlisting, please use ${configure.cxx} here instead of hardcoding the compiler name.

evanmiller commented:
Updated the code and adopted @cjones051073's suggestions regarding the run-time compiler.

This build may fail due to some recent changes in mrustc, reported here: https://github.com/thepowersgang/mrustc/issues/210

evanmiller commented:
I have isolated and fixed the recent local build failures described in the linked issue.

Updated the code and Portfile to consolidate patches and have better GCC compatibility.

I have a slight preference to wait until https://github.com/thepowersgang/mrustc/pull/192 is merged, but if CI passes this PR should be ready as-is.

evanmiller commented:
@pmetzger Will do. Still working through a GCC-related issue over here: https://github.com/thepowersgang/mrustc/issues/214

evanmiller commented:
This is OK to commit; I will address any remaining issues as they are resolved upstream.
